tests_e2e/tests/agent_update/rsm_update.py

```
@@ -0,0 +1,189 @@
+        super().__init__(context)
+        self._ssh_client = SshClient(
+            ip_address=self._context.vm_ip_address,
+            username=self._context.username,
+            private_key_file=self._context.private_key_file)
+
+    def get_ignore_error_rules(self) -> List[Dict[str, Any]]:
+        ignore_rules = [
+            #
+            # This is expected as we validate the downgrade scenario
+            #
+            # WARNING ExtHandler ExtHandler Agent WALinuxAgent-9.9.9.9 is permanently blacklisted
+            #
+            {
+                'message': r"Agent WALinuxAgent-9.9.9.9 is permanently blacklisted"
+            }
+
+        ]
+        return ignore_rules
+
+    def run(self) -> None:
+        # Allow agent to send supported feature flag
+        self._verify_agent_reported_supported_feature_flag()
+
+        log.info("*******Verifying the Agent Downgrade scenario*******")
+        self._mock_rsm_update("1.3.0.0")
+        self._check_rsm_gs("1.3.0.0")
+        self._prepare_agent()
+
+        # Verify downgrade scenario
+        self._verify_guest_agent_update("1.3.0.0")
+
+        # Verify upgrade scenario
+        log.info("*******Verifying the Agent Upgrade scenario*******")
+        self._mock_rsm_update("1.3.1.0")
+        self._check_rsm_gs("1.3.1.0")
+        self._verify_guest_agent_update("1.3.1.0")
+
+        # verify no version update. There is bug in CRP and will enable once it's fixed
+        # log.info("*******Verifying the no version update scenario*******")
+        # self._prepare_rsm_update("1.3.1.0")
+        # self._verify_guest_agent_update("1.3.1.0")
+
+    def _check_rsm_gs(self, requested_version: str) -> None:
+        # This checks if RSM GS available to the agent after we mock the rsm update request
+        output = self._ssh_client.run_command(f"rsm_goal_state.py --version {requested_version}", use_sudo=True)
+        log.info('Verifying requested version GS available to the agent \n%s', output)
```

if we don't get the RSM goal state we should raise an exception